### ISSUE TITLE
Checkpoint not working

### ISSUE THREAD

**Issue by patrickbryant1** [google/jax#12792]
### Description

Dear developers,

I have a function that applies a forward pass through a complicated network written with JAX+Haiku. I would like to checkpoint all gradients being calculated through this function in the following manner:

@functools.partial(jax.checkpoint)
def apply_fwd(params, rng, batch):
    return forward.apply(params, rng, batch)

This runs and all, but does not reduce memory consumption. Do you have any idea of why this doesn't raise errors but obviously doesn't checkpoint anything?

### What jax/jaxlib version are you using?

jaxlib>=0.3.15

### Which accelerator(s) are you using?

GPU

### Additional system info

Python3, Debian

**Comment by mattjj** [comment]
Thanks for the question!

I think we just need to use `jax.checkpoint` differently. Applying it to a function to be differentiated indeed won't result in any memory savings, as described in more detail below. Instead, we need to apply it to sub-functions.

Say we have a function `f = lambda x: f2(f1(x))` that we're interested in differentiating. That is, we're interested in evaluating `jax.grad(f)(x)`. It involves no `jax.checkpoint` decorators internally (yet).

Writing `f_grad = jax.grad(jax.checkpoint(f))` would be pretty much pointless: the computation that would result would

1. run the forward pass, discarding all intermediates;
2. immediately re-run the forward pass, saving intermediates;
3. run the backward pass, consuming the intermediates.

That is, in code we'd have spelled the computation:

```python
def f_grad(x):
  _ = f(x)  # step 1
  _, f_vjp = jax.vjp(f, x)  # step 2, entails re-running f, intermediates in f_vjp's closure
  x_bar, = f_vjp(1.0)  # step 3
  return x_bar
```

For a similar reason, the use of `jax.checkpoint` in `f_grad = jax.checkpoint(lambda x: jax.checkpoint(f2)(f1(x)))` would also be pointless, since we'd be doing:

1. run the forward pass of `f1`, saving intermediates;
2. run the forward pass of `f2`, discarding all intermediates;
3. re-run the forward pass of `f2`, saving intermediates;
4. run the backward pass of `f2`, consuming the relevant intermediates;
5. run the backward pass of `f1`, consuming the relevant intermediates.

That is, we'd have spelled the computation

```python
def f_grad(x):
  y, f1_vjp = jax.vjp(f1, x)  # step 1
  _ = f2(y)  # step 2
  _, f2_vjp = jax.vjp(f2, y)  # step 3, entails re-running f2
  y_bar, = f2_vjp(1.0)  # step 4
  x_bar, = f1_vjp(y_bar)  # step 5
  return x_bar
```

To get any savings using `jax.checkpoint`, we need something like `f_grad = jax.grad(lambda x: f2(jax.checkpoint(f1)(x)))`, since that would:

1. run the forward pass of `f1`, discarding intermediates;
2. run the forward pass of `f2`, saving intermediates;
3. run the backward pass of `f2`, consuming the relevant intermediates;
4. re-run the forward pass of `f1`, saving intermediates;
5. run the backward pass of `f1`, consuming the relevant intermediates.

That is, we'd finally have:

```python
def f_grad(x):
  y = f1(x)  # step 1
  _, f2_vjp = jax.vjp(f2, y)  # step 2
  y_bar, = f2_vjp(1.0)  # step 3
  _, f1_vjp = jax.vjp(f1, x)  # step 4
  x_bar, = f1_vjp(y_bar)  # step 5
  return x_bar
```

More generally, if we had a chain composition of more functions, like `g = lambda x: g3(g2(g1(x)))`, and we were interested in evaluating `jax.grad(g)`, we could say that

1. we shouldn't apply `jax.checkpoint` to the whole function to be differentiated, i.e. `jax.checkpoint(g)` would not save memory;
2. we shouldn't apply `jax.checkpoint` to the last function, i.e. `jax.checkpoint(g3)` would not save memory;
3. applying `jax.checkpoint` to `g1`, `g2`, or their composition `lambda x: g2(g1(x))` would all make sense and express different memory/compute tradeoffs.

A bit more concretely, in your example you probably don't want to apply `jax.checkpoint` to `apply_fwd`, but rather to individual layers or compositions of inner layers.

I added the 'documentation' tag because I'd like to improve the documentation around this.

What do you think?

**Comment by mattjj** [comment]
Actually, there is another option to decorating subroutine functions, though it's totally undocumented: you can use a top-level `jax.checkpoint` with a policy function which describes what should be considered saveable, e.g.

```python
@functools.partial(jax.checkpoint, policy=jax.checkpoint_policies.checkpoint_dots)
def apply_fwd(params, rng, batch):
  return forward.apply(params, rng, batch)
```

That code should result in memory savings because instead of having to recompute absolutely everything on the backward pass, we'll save the results of intermediate dots (i.e. matmuls, batched matmuls, etc) and only recompute things from there as needed.

The list of policies is [here](https://github.com/google/jax/blob/d0e33b7ce08e5e27969a66fcc61e32e0507e847e/jax/_src/ad_checkpoint.py#L105).

**Comment by patrickbryant1** [comment]
Dear Matthew, 
Thank you for the thorough (and great) answer. I understand why the checkpoint doesn't work now and I suspected that I would have to apply the checkpointing to each individual function. The problem for me is that all my functions are really Haiku and thus only "transformed functions" and I suspect this changes things(?).  It is also not really possible for me to wrap all individual functions in a checkpoint since too much is really going on to make this tractable for me.

The suggestion of wrapping apply_fwd in @functools.partial(jax.checkpoint, policy=jax.checkpoint_policies.checkpoint_dots) is great, but sadly results in a higher memory consumption (11.5 times higher). I am not sure what is going on here, but the error message only suggests OOM. 

The idea to improve the documentation sounds great. I would be happy to help to improve the documentation since I think checkpointing is a very useful feature.

**Comment by mattjj** [comment]
I'm not sure what's up with the higher memory usage. You could experiment with the `jax.ad_checkpoint.print_saved_residuals` helper function, which tells you fairly precisely which which residuals will be saved when you differentiate a function:

```
$ cat -n patrick3.py
     1  import jax
     2  import jax.numpy as jnp
     3
     4
     5  def f(x):
     6    y = jnp.sin(x)
     7    z = jnp.sin(y)
     8    return z
     9
    10  jax.ad_checkpoint.print_saved_residuals(f, jnp.arange(3.))
$ python patrick3.py
f32[3] from /usr/local/google/home/mattjj/packages/jax/patrick3.py:6 (f)
f32[3] from /usr/local/google/home/mattjj/packages/jax/patrick3.py:7 (f)
```

You can explore how adding different checkpoint policies affects these. But I'm not sure how well it scales to complex networks.

Should we close this issue, since the original question is answered?

**Comment by patrickbryant1** [comment]
Ok, I understand. Thanks for the help - let's close it.
